chia/rpc/full_node_rpc_api.py

```diff
@@ -577,3 +597,344 @@ async def get_mempool_item_by_tx_id(self, request: Dict) -> Optional[Dict]:
+            if last_solution is None:
+                raise ValueError("Invalid launcher solution")
+            delay_time, delay_puzzle_hash = get_delayed_puz_info_from_launcher_spend(last_solution)
+            saved_state = solution_to_extra_data(last_solution)
+            if saved_state is None:
+                raise ValueError("Invalid launcher coin")
+        else:
+            farmer_record = request["farmer_record"]
+            if "singleton_tip" not in farmer_record:
+                raise ValueError("farmer_record.singleton_tip not in request")
+            if "singleton_tip_state" not in farmer_record:
+                raise ValueError("farmer_record.singleton_tip_state not in request")
+            if "delay_time" not in farmer_record:
+                raise ValueError("farmer_record.delay_time not in request")
+            if "delay_puzzle_hash" not in farmer_record:
+                raise ValueError("farmer_record.delay_puzzle_hash not in request")
+
+            last_solution = CoinSolution.from_json_dict(farmer_record["singleton_tip"])
+            saved_state = PoolState.from_json_dict(farmer_record["singleton_tip_state"])
+            if last_solution is None or saved_state is None:
+                raise ValueError("Invalid singleton_tip or singleton_tip_state")
+
+            delay_time = uint64(farmer_record["delay_time"])
+            delay_puzzle_hash = bytes32(hexstr_to_bytes(farmer_record["delay_puzzle_hash"]))
+
+        assert last_solution is not None and saved_state is not None
+        saved_coin_record: Optional[CoinRecord] = await self.service.blockchain.coin_store.get_coin_record(last_solution.coin.name())
+        if saved_coin_record is None:
+            raise ValueError(f"Not found last solution coin 0x{last_solution.coin.name()}")
+        saved_solution: CoinSolution = last_solution # Last coin solution that is buried in the blockchain, for this singleton
+        last_not_none_state: PoolState = saved_state # Current state of the singleton
+
+        while True:
+            next_coin: Optional[Coin] = get_most_recent_singleton_coin_from_coin_solution(last_solution)
+            if next_coin is None:
+                raise ValueError("The singleton is invalid")
+            next_coin_record: Optional[CoinRecord] = await self.service.blockchain.coin_store.get_coin_record(next_coin.name())
+            if next_coin_record is None:
+                raise ValueError(f"Not found coin 0x{next_coin.name()}")
+            if not next_coin_record.spent:
+                if not self._validate_pool_puzzle_hash(
+                    launcher_id,
+                    delay_puzzle_hash,
+                    delay_time,
+                    last_not_none_state,
+                    next_coin_record.coin.puzzle_hash,
+                    self.service.constants.GENESIS_CHALLENGE,
+                ):
+                    raise ValueError(f"Invalid singleton puzzle_hash for 0x{launcher_id}")
+                break
```

Comment:
this whole block is copy pasted (from line 729). It should probably be turned into a function